crypto/bio/bss_file.c

```diff
@@ -89,6 +89,10 @@
 
 # if !defined(OPENSSL_NO_STDIO)
 
+#  if defined(OPENSSL_SYS_MSDOS) && !defined(OPENSSL_SYS_WINDOWS)
+#   include <libc/unconst.h>
+static void dosify_filename(const char *filename);
+#  endif
```

Comment:
I'd argue that this is definition of bad idea. First of all #if condition is totally misleading. Because it's really something specific to DJGPP, which is effectively a POSIX-like environment which *happens* to run under DPMI. It's not really DOS and mentioning that it's not Windows either doesn't make it clearer. This is really case when compiler pre-define such as __DJGPP__ does much better job.

Secondly [and more importantly!] breaking const contract is hardly appropriate. It might work in specific situation (like our test suite), but it's error-prone in general case. And real trouble is that in latter case it would *have to* be treated as bug. Which means that one effectively introduces a bug here. Which "general case"? Well, on normal OSes attempt to write to const might actually result in SEGV. I don't know if DJGPP has notion of read-only pages, but even if it doesn't, programmer who passes string to function declared as accepting const one is *entitled* to expect that it doesn't change. Let's say it's passed later to some other program component and original format is the one to be used... Right thing to do is to allocate temporary buffer (on stack?), copy filename, "dosify" it, fopen and return file descriptor. *And* to minimize clobber in common source it would probably be appropriate to implement it in separate file, and override calls for fopen with say __djgpp_fopen by means of some pre-processor magic. Extra -D in config might be appropriate, and separate file can be injected through say uplink_aux_src? Note that in such case one would probably be able to remove all these #ifs, which is desirable, isn't it? Minimizing #if spaghetti that is.